src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+            {
+                foreach (BoundSwitchLabel label in section.SwitchLabels)
+                {
+                    if (label.Syntax.Kind() != SyntaxKind.DefaultSwitchLabel)
+                    {
+                        if (label.Pattern.HasErrors)
+                        {
+                            return;
+                        }
+
+                        existingCases.Add(builder.MakeTestsForPattern(++index, label.Syntax, rootIdentifier, label.Pattern, label.WhenClause, label.Label));
+                    }
+                }
+            }
+
+            int patternIndex = 0;
+            foreach (BoundSwitchSection section in switchSections)
+            {
+                foreach (BoundSwitchLabel label in section.SwitchLabels)
+                {
+                    if (label.Syntax.Kind() != SyntaxKind.DefaultSwitchLabel)
+                    {
+                        CheckOrAndAndReachability(existingCases, patternIndex, label.Pattern, builder, rootIdentifier, defaultLabel, syntax, diagnostics);
+                        patternIndex++;
+                    }
+                }
+            }
+
+            existingCases.Free();
+        }
+
+        private static void CheckOrAndAndReachability(
+          ArrayBuilder<StateForCase> previousCases,
+          int patternIndex,
+          BoundPattern pattern,
+          DecisionDagBuilder builder,
+          BoundDagTemp rootIdentifier,
+          LabelSymbol defaultLabel,
+          SyntaxNode syntax,
+          BindingDiagnosticBag diagnostics,
+          bool isSwitchExpression = false)
+        {
+            CheckOrReachability(previousCases, patternIndex, pattern,
+                builder, rootIdentifier, defaultLabel, syntax, diagnostics, ignoreDefaultLabel: isSwitchExpression);
+
+            CheckOrReachability(previousCases, patternIndex, MoveNotPatternsDownRewriter.MakeNegatedPattern(pattern),
+                builder, rootIdentifier, defaultLabel, syntax, diagnostics, ignoreDefaultLabel: isSwitchExpression);
+        }
+
+        private static void CheckOrReachability(
+            ArrayBuilder<StateForCase> previousCases,
+            int patternIndex,
+            BoundPattern pattern,
+            DecisionDagBuilder builder,
+            BoundDagTemp rootIdentifier,
+            LabelSymbol defaultLabel,
+            SyntaxNode syntax,
+            BindingDiagnosticBag diagnostics,
+            bool ignoreDefaultLabel = false)
+        {
+            var normalizedPattern = MoveNotPatternsDownRewriter.Rewrite(pattern);
+
+            SetsOfOrCases setOfOrCases = RewriteToSetsOfOrCases(normalizedPattern);
+            if (setOfOrCases.IsDefault)
+            {
+                return;
+            }
+
+            // We construct a DAG and analyze reachability of branches once per `or` sequence
+            Debug.Assert(!setOfOrCases.IsDefault);
+            foreach (OrCases orCases in setOfOrCases.Set)
```

Comment:
>foreach (OrCases orCases in setOfOrCases.Set) [](http://example.com/codeflow?start=12&length=45)

I think we should be possible to avoid realizing all sets of cases in memory before analyzing them

Reply:
I agree it's possible, but not planning to make such a change at the moment.

Reply:
> but not planning to make such a change at the moment.

Let's discuss this offline.